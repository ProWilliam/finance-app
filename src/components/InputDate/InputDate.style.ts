import { StyleSheet } from 'react-native';
import theme from '../../theme/theme';

const StyleInputDate = StyleSheet.create({
  container: {
    alignItems: 'center',
    backgroundColor: theme.light.tertiary,
    borderRadius: 5,
    color: theme.light.primary
  },
  textInput: {
    color: theme.light.quaternary,
    fontSize: theme.size.fontLittle,
    padding: 10,
  },
  inputWeb: {
    height: 50,
    paddingLeft: 20,
    paddingRight: 20,
    borderColor: theme.light.border,
    borderWidth: 1,
    borderRadius: 5,
  }
});


export default StyleInputDate