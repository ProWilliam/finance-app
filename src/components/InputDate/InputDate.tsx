import { useState } from 'react';
import { View, Platform, Text, TouchableOpacity } from 'react-native';
import DateTimePickerModal from "react-native-modal-datetime-picker";
import { FormatingDate } from '../../utils/FormatingDate';

// Type and Styles.
import styles from './InputDate.style';
import DateProps from '../../types/components/InputDate.types';



const InputDate: React.FC<DateProps> = ({ title, keyValue, onChangeForm, value }) => {

  const [dateWeb, setDateWeb] = useState('2026-08-22')
  
  const [isDatePickerVisible, setDatePickerVisibility] = useState(false);
  const [buttonTitle, setButtonTitle] = useState<string>(title);

  const showDatePicker = () => {
    setDatePickerVisibility(true);
  };

  const hideDatePicker = () => {
    setDatePickerVisibility(false);
  };

  const onDateWeb = (event: any) => {
    const dateFormWeb = event.target.value;

    onChangeForm(keyValue, dateFormWeb);
    setDateWeb(dateFormWeb)
  }

  const handleConfirm = (date: any) => {
    const dateFormat = FormatingDate(date);

    onChangeForm(keyValue, dateFormat);
    setButtonTitle(dateFormat);
    hideDatePicker();
  };

  return (
    <View>
      { Platform.OS == 'web' && (
        <input
          style={styles.inputWeb}
          type="date"
          value={dateWeb}
          onChange={onDateWeb}
        />
      )}
      { Platform.OS != 'web' && (
        <View>
          <TouchableOpacity
            style={styles.container}
            onPress={showDatePicker}
          >
            <Text
              style={styles.textInput}
            >
              {value || buttonTitle}
            </Text>
          </TouchableOpacity>
          <DateTimePickerModal
            isVisible={isDatePickerVisible}
            mode="date"
            onConfirm={handleConfirm}
            onCancel={hideDatePicker}
          />
        </View>
        
      )}
    </View>
  );
};

export default InputDate;