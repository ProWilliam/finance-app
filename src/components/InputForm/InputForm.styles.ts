import { StyleSheet } from 'react-native';
import theme from '../../theme/theme';


const StylesInput = StyleSheet.create({
  containerInput:{
    marginBottom: 5
  },
  input: {
    width: `${theme.withScreen.big}%`,
    height: 50,
    borderWidth: 1,
    borderRadius: 5,
    paddingLeft: 20,
    fontSize: theme.size.fontLittle,
    
  },
  textError: {
    marginTop: 5,
    fontSize: 14,
    color: theme.light.sentary,
  },
  borderError: {
    borderColor: theme.light.sentary,
  },
  border: {
    borderColor: theme.light.tertiary,
  },
  colorPlaceHolder: {
    color: theme.light.quintary,
  }
})

export default StylesInput;