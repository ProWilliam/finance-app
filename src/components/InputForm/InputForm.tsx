import React, { useState } from 'react';
import { TextInput, View, Text, KeyboardTypeOptions } from 'react-native';

// Components
import InputDate from '../InputDate/InputDate';

// Styles and Types
import styles from './InputForm.styles';
import InputFormProps from '../../types/components/InputForm.types';


const InputForm: React.FC<InputFormProps> = ({placeHolder, keySelect, errorMessage, onChange, type, value }) => {

  const [error, setError] = useState<string | null>(errorMessage || null);
  const [text, setText] = useState<string>('');

  const validatorText = () => {
    if(text.trim() === ''){
      setError('Este campo es obligatorio!');
    } else {
      setError(null);
    }
  }

  const handleChangeText = (value: string) => {

    setText(value);
    onChange(keySelect, value);

    if (value.trim() !== '') {
      setError(null);
    }else {
      setError('Este campo es obligatorio!');
    }
  }

  // Change of styles for action
  const inputStyles = [
    styles.input,
    error !== null
    ? styles.borderError 
    : styles.border,
  ]

  return (
    <View style={styles.containerInput}>
      {
        type ==='date'
        ? <InputDate 
            title={placeHolder}
            keyValue={keySelect}
            onChangeForm={onChange}
            value={value}
          />
        :
        <TextInput 
          style={inputStyles} 
          placeholder={placeHolder}
          placeholderTextColor={styles.colorPlaceHolder.color}
          onBlur={validatorText}
          onChangeText={(text) => handleChangeText(text)}
          value={value}
          blurOnSubmit 
          keyboardType={type as KeyboardTypeOptions}
          testID='error-message'
        /> 

      }      

      { error && <Text testID='error-message' style={styles.textError}>{error}</Text> }
    </View>
  )
}

export default InputForm;