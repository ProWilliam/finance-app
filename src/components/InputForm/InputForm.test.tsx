import React from 'react';
import renderer, { act } from 'react-test-renderer';
import { TextInput } from 'react-native';

// Components
import InputDate from '../InputDate/InputDate';
import InputForm from './InputForm';
import styles from './InputForm.styles';


describe('<InputForm />', () => {

  let onChangeMock: jest.Mock;

  beforeEach(() => {
    onChangeMock = jest.fn();
  });

  it('renders correctly with TextInput', () => {
    const tree = renderer.create(
      <InputForm
        placeHolder="Numero de ID"
        keySelect="id"
        errorMessage="Campo obligatorio"
        onChange={onChangeMock}
        type="numeric"
        value="1"
      />
    ).toJSON();
    
    expect(tree).toMatchSnapshot();
  });

  it('should render TextInput and handle text changes', () => {

    const tree = renderer.create(
      <InputForm
        placeHolder="Tarjeta Crédito"
        keySelect="name"
        errorMessage=""
        onChange={onChangeMock}
        type="default"
        value=""
      />
    );

    const input = tree.root.findByType(TextInput);
    expect(input.props.placeholder).toBe('Tarjeta Crédito');
    
    // Simulate text change
    act(() => {
      input.props.onChangeText('New text');
    });

    // Check if onChange is called with the correct parameters
    expect(onChangeMock).toHaveBeenCalledWith('name', 'New text');
    
    // Verify that error is cleared
    const errorText = tree.root.findByProps({ testID: 'error-message' });
    expect(errorText.props.children).toBeUndefined();
  });
  
  it('should show error message when input is blurred with empty value', () => {
    const tree  = renderer.create(
      <InputForm
        placeHolder="Descripción"
        keySelect="description"
        errorMessage=""
        onChange={onChangeMock}
        type="default"
        value=""
      />
    );

    const input = tree.root.findByType(TextInput);
    
    // Simulate blur event
    act(() => {
      input.props.onBlur();
    });

    // Verify error message is shown
    const errorText = tree.root.findByProps({ style: styles.textError });
    expect(errorText.props.children).toBe('Este campo es obligatorio!');
  });

  it('calls onChange handler with correct value on text change and date ID is numeric', () => {
    const tree = renderer.create(
      <InputForm
        placeHolder="ID"
        keySelect="id"
        errorMessage="Campo obligatorio"
        onChange={onChangeMock}
        type="numeric"
        value=""
      />
    );

    // Find the TextInput component
    const textInput = tree.root.findByType(TextInput);

    act(() => {
      textInput.props.onChangeText('1');
    });

    // Verify the Key and Value.
    expect(onChangeMock).toHaveBeenCalledWith('id', '1');

    // Verify the numeric conversion
    const args = onChangeMock.mock.calls[0][1]; // Get the argument passed to onChangeMock
    const numericValue = Number(args);

    expect(typeof numericValue).toBe('number');
    expect(isNaN(numericValue)).toBe(false); // Ensure it is not NaN

  });

  it('should render InputDate component for type="date"', () => {
    const tree = renderer.create(
      <InputForm
        placeHolder="22/02/2023"
        keySelect="date_release"
        errorMessage=""
        onChange={onChangeMock}
        type="date"
        value="2024-08-10"
      />
    );

    // Verify InputDate component is rendered
    const inputDate = tree.root.findByType(InputDate);
    expect(inputDate.props.title).toBe('22/02/2023');
    expect(inputDate.props.keyValue).toBe('date_release');
    expect(inputDate.props.value).toBe('2024-08-10');
  });


});