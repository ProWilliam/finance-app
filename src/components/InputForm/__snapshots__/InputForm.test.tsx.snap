// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InputForm /> renders correctly with TextInput 1`] = `
<View
  style={
    {
      "marginBottom": 5,
    }
  }
>
  <TextInput
    blurOnSubmit={true}
    keyboardType="numeric"
    onBlur={[Function]}
    onChangeText={[Function]}
    placeholder="Numero de ID"
    placeholderTextColor="#6a6a6a"
    style={
      [
        {
          "borderRadius": 5,
          "borderWidth": 1,
          "fontSize": 24,
          "height": 50,
          "paddingLeft": 20,
          "width": "100%",
        },
        {
          "borderColor": "#F70000",
        },
      ]
    }
    testID="error-message"
    value="1"
  />
  <Text
    style={
      {
        "color": "#F70000",
        "fontSize": 14,
        "marginTop": 5,
      }
    }
    testID="error-message"
  >
    Campo obligatorio
  </Text>
</View>
`;
